subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/IgnoreResourceFilter.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.changedetection.state;
+
+import com.google.common.collect.ImmutableSet;
+import org.gradle.api.file.RelativePath;
+import org.gradle.api.internal.file.pattern.PatternMatcherFactory;
+import org.gradle.api.specs.Spec;
+import org.gradle.caching.internal.BuildCacheHasher;
+
+import java.util.Set;
+
+public class IgnoreResourceFilter implements MetadataFilter {
+    private final Set<String> ignores;
+    private final ImmutableSet<Spec<RelativePath>> ignoreSpecs;
+
+    public IgnoreResourceFilter(Set<String> ignores) {
+        this.ignores = ImmutableSet.copyOf(ignores);
+        ImmutableSet.Builder<Spec<RelativePath>> builder = ImmutableSet.builder();
+        for (String ignore : ignores) {
+            Spec<RelativePath> matcher = PatternMatcherFactory.getPatternMatcher(false, true, ignore);
+            builder.add(matcher);
+        }
+        this.ignoreSpecs = builder.build();
```

Comment:
Wrapping each `PathMatcher` in a `Spec` and calling through repeatedly seems like a lot of waste. `Spec`s are to capture logic that needs to be used someplace else. We are capturing logic that we want to use in-place here, so we are misusing `Spec`.

I don't see a problem with calling `getSegments()`, nor with adding an overload to `PathMatcher` to accept `RelativePath`s.